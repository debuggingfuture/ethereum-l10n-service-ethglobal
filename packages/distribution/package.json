{
  "name": "@repo/distribution",
  "version": "0.0.0",
  "private": "true",
  "module": "esnext",
  "scripts": {
    "lint": "eslint . --max-warnings 0",
    "test": "vitest"
  },
  "devDependencies": {
    "@next/eslint-plugin-next": "^14.1.4",
    "@repo/eslint-config": "workspace:*",
    "@repo/typescript-config": "workspace:*",
    "@types/eslint": "^8.56.5",
    "@types/js-levenshtein": "^1.1.3",
    "@types/lodash": "^4.17.0",
    "@types/node": "^20.11.24",
    "eslint": "^8.57.0",
    "eslint-config-prettier": "^9.1.0",
    "eslint-plugin-jest": "^28.0.0",
    "eslint-plugin-prettier": "^5.0.0",
    "fast-check": "^3.17.1",
    "prettier": "^3.2.5",
    "typescript": "^5.4.4",
    "vitest": "^1.5.0"
  },
  "dependencies": {
    "decimal.js": "^10.4.3",
    "js-levenshtein": "^1.1.6",
    "lodash": "^4.17.21"
  }
}
