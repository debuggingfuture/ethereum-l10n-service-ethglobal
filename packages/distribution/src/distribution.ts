import {
  each,flatten, flow, fromPairs, get, groupBy, invoke,
  map, mapValues, sum, sumBy, sortBy, tap, toPairs, values
} from 'lodash/fp';
import rawMap from 'lodash/map';
import Decimal from 'decimal.js';

import levenshtein from 'js-levenshtein';

export type WalletAddress = string;

// 0.1 represents 10%
export type AllocationRatio = Decimal;

export type Distribution = [WalletAddress, AllocationRatio];

export type TranslationInput = {
  contributorAddress: WalletAddress,
  lineId: string,
  translatedLineId: string,
  contributedAtMs: bigint,
  translatedText: string,
  // assuming value [1, 100]
  score: number | Decimal
};

export type Translation = TranslationInput & {
  score: Decimal
};

export type WithEditDistance = Translation & {
  // between 0 and EDIT_DISTANCE_MULTIPLIER_SCALE
  editDistanceMultiplier: number
};

export type WithDecay = Translation & {
  // between 0 and 1
  decayMultiplier: number
};

export type WithAdjustedScore = Translation & {
  adjustedScore: Decimal
};

// Scale of effect of edit distance.
export const EDIT_DISTANCE_MULTIPLIER_SCALE = 1

// 365 days - meaning contribution after 365 days the decay multiplier = 0
export const DEFAULT_DECAY_THRESHOLD = 365 * 24 * 60 * 60 * 1000;

export const sumDecimal = (them: Decimal[]) => them.length > 0 ? Decimal.sum(...them) : new Decimal(0)
export const sumByDecimal = (key: string) => flow(map(get(key)), sumDecimal)

/**
 * @returns A flattened list of distribution. The sum of all AllocationRatio <= 1.
 */
export function calculateDistribution(inputs: Translation[]): Distribution[] {
  return flow(
    map((it: TranslationInput) => ({ ...it, score: new Decimal(it.score) })),
    each(validate),
    createAdjustedScores(),
    sumByWallet,
    normalizeScore,
  )(inputs)
}

export function validate(input: Translation): void {
  const { score } = input

  if (score.lessThan(1) || score.greaterThan(100)) {
    throw new Error('score must be [1, 100]')
  }
}

/**
 * Injects `adjustedScore` to each item.
 */
export const createAdjustedScores = (
  injectEditDistanceMultiplier = createEditDistanceMultiplier(),
  injectDecayMultiplier = createDecayMultiplier()
) => flow(
  groupBy('lineId'),
  mapValues(sortBy('contributedAtMs')),
  mapValues(injectEditDistanceMultiplier),
  mapValues(injectDecayMultiplier),
  mapValues(map((curr: Translation & WithEditDistance & WithDecay): Translation & WithAdjustedScore => ({
    ...curr,
    adjustedScore: curr.score.times(curr.editDistanceMultiplier).times(curr.decayMultiplier)
  }))),
  values,
  flatten,
)

/**
 * Group and sum scores by wallet address.
 */
export const sumByWallet = flow(
  groupBy('contributorAddress'),
  mapValues(sumByDecimal('adjustedScore')),
  toPairs,
)

/**
 * Normalize adjusted scores into percentage value.
 */
export const normalizeScore = (scores: [string, number][]) => {
  const total = flow(map(get(1)), sumDecimal)(scores)
  return flow(
    fromPairs,
    mapValues((it: number) => new Decimal(it)),
    mapValues((it: Decimal) => it.dividedBy(total)),
    toPairs
  )(scores)
}

/**
 * Calculates multiplier based on edit distance / text length.
 */
export const createEditDistanceMultiplier = (scale: number = EDIT_DISTANCE_MULTIPLIER_SCALE) =>
  (them: Translation[] ): Translation[] =>
    rawMap(them, (curr, i, array): Translation & WithEditDistance => {
      if (i === 0) {
        return { ...curr, editDistanceMultiplier: 1 };
      }

      const last = array[i - 1]!
      const editDistance = levenshtein(last.translatedText, curr.translatedText)
      const editDistanceMultiplier = scale * editDistance / last.translatedText.length

      return { ...curr, editDistanceMultiplier }
    }
  )

export const createDecayMultiplier = (threshold: number = DEFAULT_DECAY_THRESHOLD) =>
  (them: Translation[]) =>
    rawMap(them, (curr, i, array): Translation & WithDecay => {
      if (i === 0) {
        return { ...curr, decayMultiplier: 1 };
      }

      const timeDiff = curr.contributedAtMs - array[0]!.contributedAtMs;
      const decayMultiplier = 1 - (timeDiff <= Number.MAX_SAFE_INTEGER ? Number(timeDiff) : Number.MAX_SAFE_INTEGER) / threshold

      return {
        ...curr,
        decayMultiplier: decayMultiplier > 0 ? Number(decayMultiplier) : 0
      };
    }
  )

export const debugLog = tap(it => console.debug('[DEBUG] ==> ', it))
