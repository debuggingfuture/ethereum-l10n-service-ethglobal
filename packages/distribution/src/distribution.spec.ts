import { describe, test, expect } from 'vitest';
import fc from 'fast-check';
import {
  each,flatten, flow, get, groupBy, map, mapValues,
  sortBy, tap, toPairs, values
} from 'lodash/fp';
import {
  type Translation,
  type WithAdjustedScore,
  calculateDistribution,
  createAdjustedScores,
  createDecayMultiplier,
  createEditDistanceMultiplier,
  sumDecimal,
  debugLog,
} from "./distribution";
import { tranlsationListArbitary, translationsFixture } from './distribution.fixture';
import Decimal from 'decimal.js';

describe('createEditDistanceMultiplier()', () => {
  const injectEditDistanceMultiplier = createEditDistanceMultiplier()

  test('it adds "editDistance" to tranlsation inputs', () => {
    expect(injectEditDistanceMultiplier(translationsFixture)).toMatchSnapshot()
  })
})

describe('decayMultiplier()', () => {
  const injectDecayMultiplier = createDecayMultiplier()

  test('it adds "decayMultiplier" to tranlsation objects', () => {
    expect(injectDecayMultiplier(translationsFixture)).toMatchSnapshot()
  })
})

describe('createAdjustedScores()', () => {
  const injectAdjustedScores = createAdjustedScores()

  describe('Properties', () => {
    test('adjustedScore >= 0', () => {
      fc.assert(
        fc.property(tranlsationListArbitary(), (inputs) => {
          const outputs = injectAdjustedScores(inputs)
          outputs.forEach((it: Translation & WithAdjustedScore) => {
            expect(it.adjustedScore.greaterThanOrEqualTo(0)).toBeTruthy()
          })
        })
      )
    })
  })

  test('it adds "adjustedScores" to tranlsation inputs', () => {
    expect(injectAdjustedScores(translationsFixture)).toMatchSnapshot()
  })
})

describe('calculateDistribution()', () => {
  test('it returns [[address, distributionPercentage]]', () => {
    expect(calculateDistribution(translationsFixture)).toMatchSnapshot()
  })

  describe('Properties', () => {
    test('total distribution in range [0, 1]', () => {
      fc.assert(
        fc.property(tranlsationListArbitary(), (inputs) => {
          const sum: Decimal = flow(
            map(get(1)),
            sumDecimal,
          )(calculateDistribution(inputs))

          expect(sum.lessThanOrEqualTo(1)).toBeTruthy()
        })
      )
    })
  })
})
