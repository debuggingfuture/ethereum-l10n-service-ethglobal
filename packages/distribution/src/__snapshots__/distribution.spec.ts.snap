// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`calculateDistribution() > it returns [[address, distributionPercentage]] 1`] = `
[
  [
    "0x1234",
    "0.61351474661127737537",
  ],
  [
    "0x5678",
    "0.11011801398395041019",
  ],
  [
    "0x9abc",
    "0.20450485068900913232",
  ],
  [
    "0xdef0",
    "0.032534769758950351442",
  ],
  [
    "0x2234",
    "0.039327618956812730678",
  ],
]
`;

exports[`createAdjustedScores() > it adds "adjustedScores" to tranlsation inputs 1`] = `
[
  {
    "adjustedScore": "10",
    "contributedAtMs": 1618000000000n,
    "contributorAddress": "0x1234",
    "decayMultiplier": 1,
    "editDistanceMultiplier": 1,
    "lineId": "line1",
    "score": "10",
    "translatedLineId": "translatedLine1",
    "translatedText": "Hello, world!",
  },
  {
    "adjustedScore": "10.769229061780429538",
    "contributedAtMs": 1618000005000n,
    "contributorAddress": "0x5678",
    "decayMultiplier": 0.99999984145104,
    "editDistanceMultiplier": 0.5384615384615384,
    "lineId": "line1",
    "score": "20",
    "translatedLineId": "translatedLine2",
    "translatedText": "Goodbye, world!",
  },
  {
    "adjustedScore": "19.999993658041602",
    "contributedAtMs": 1618000010000n,
    "contributorAddress": "0x9abc",
    "decayMultiplier": 0.9999996829020802,
    "editDistanceMultiplier": 0.6666666666666666,
    "lineId": "line1",
    "score": "30",
    "translatedLineId": "translatedLine3",
    "translatedText": "Hello again, world!",
  },
  {
    "adjustedScore": "50",
    "contributedAtMs": 1618000100000n,
    "contributorAddress": "0x1234",
    "decayMultiplier": 1,
    "editDistanceMultiplier": 1,
    "lineId": "line2",
    "score": "50",
    "translatedLineId": "translatedLine4",
    "translatedText": "Hello, Bob!",
  },
  {
    "adjustedScore": "3.1818080923389146818",
    "contributedAtMs": 1618000200000n,
    "contributorAddress": "0xdef0",
    "decayMultiplier": 0.9999968290208017,
    "editDistanceMultiplier": 0.6363636363636364,
    "lineId": "line2",
    "score": "5",
    "translatedLineId": "translatedLine5",
    "translatedText": "Goodbye, Bob!",
  },
  {
    "adjustedScore": "3.8461294540061665",
    "contributedAtMs": 1618000300000n,
    "contributorAddress": "0x2234",
    "decayMultiplier": 0.9999936580416032,
    "editDistanceMultiplier": 0.7692307692307693,
    "lineId": "line2",
    "score": "5",
    "translatedLineId": "translatedLine6",
    "translatedText": "Hello again, Bob!",
  },
  {
    "adjustedScore": "0",
    "contributedAtMs": 1649536100000n,
    "contributorAddress": "0x1234",
    "decayMultiplier": 0,
    "editDistanceMultiplier": 0.47058823529411764,
    "lineId": "line2",
    "score": "2",
    "translatedLineId": "translatedLine7",
    "translatedText": "Hello, Bobby!",
  },
]
`;

exports[`createEditDistanceMultiplier() > it adds "editDistance" to tranlsation inputs 1`] = `
[
  {
    "contributedAtMs": 1618000000000n,
    "contributorAddress": "0x1234",
    "editDistanceMultiplier": 1,
    "lineId": "line1",
    "score": "10",
    "translatedLineId": "translatedLine1",
    "translatedText": "Hello, world!",
  },
  {
    "contributedAtMs": 1618000005000n,
    "contributorAddress": "0x5678",
    "editDistanceMultiplier": 0.5384615384615384,
    "lineId": "line1",
    "score": "20",
    "translatedLineId": "translatedLine2",
    "translatedText": "Goodbye, world!",
  },
  {
    "contributedAtMs": 1618000010000n,
    "contributorAddress": "0x9abc",
    "editDistanceMultiplier": 0.6666666666666666,
    "lineId": "line1",
    "score": "30",
    "translatedLineId": "translatedLine3",
    "translatedText": "Hello again, world!",
  },
  {
    "contributedAtMs": 1618000200000n,
    "contributorAddress": "0xdef0",
    "editDistanceMultiplier": 0.7368421052631579,
    "lineId": "line2",
    "score": "5",
    "translatedLineId": "translatedLine5",
    "translatedText": "Goodbye, Bob!",
  },
  {
    "contributedAtMs": 1618000300000n,
    "contributorAddress": "0x2234",
    "editDistanceMultiplier": 0.7692307692307693,
    "lineId": "line2",
    "score": "5",
    "translatedLineId": "translatedLine6",
    "translatedText": "Hello again, Bob!",
  },
  {
    "contributedAtMs": 1618000100000n,
    "contributorAddress": "0x1234",
    "editDistanceMultiplier": 0.35294117647058826,
    "lineId": "line2",
    "score": "50",
    "translatedLineId": "translatedLine4",
    "translatedText": "Hello, Bob!",
  },
  {
    "contributedAtMs": 1649536100000n,
    "contributorAddress": "0x1234",
    "editDistanceMultiplier": 0.18181818181818182,
    "lineId": "line2",
    "score": "2",
    "translatedLineId": "translatedLine7",
    "translatedText": "Hello, Bobby!",
  },
]
`;

exports[`decayMultiplier() > it adds "decayMultiplier" to tranlsation objects 1`] = `
[
  {
    "contributedAtMs": 1618000000000n,
    "contributorAddress": "0x1234",
    "decayMultiplier": 1,
    "lineId": "line1",
    "score": "10",
    "translatedLineId": "translatedLine1",
    "translatedText": "Hello, world!",
  },
  {
    "contributedAtMs": 1618000005000n,
    "contributorAddress": "0x5678",
    "decayMultiplier": 0.99999984145104,
    "lineId": "line1",
    "score": "20",
    "translatedLineId": "translatedLine2",
    "translatedText": "Goodbye, world!",
  },
  {
    "contributedAtMs": 1618000010000n,
    "contributorAddress": "0x9abc",
    "decayMultiplier": 0.9999996829020802,
    "lineId": "line1",
    "score": "30",
    "translatedLineId": "translatedLine3",
    "translatedText": "Hello again, world!",
  },
  {
    "contributedAtMs": 1618000200000n,
    "contributorAddress": "0xdef0",
    "decayMultiplier": 0.9999936580416032,
    "lineId": "line2",
    "score": "5",
    "translatedLineId": "translatedLine5",
    "translatedText": "Goodbye, Bob!",
  },
  {
    "contributedAtMs": 1618000300000n,
    "contributorAddress": "0x2234",
    "decayMultiplier": 0.9999904870624049,
    "lineId": "line2",
    "score": "5",
    "translatedLineId": "translatedLine6",
    "translatedText": "Hello again, Bob!",
  },
  {
    "contributedAtMs": 1618000100000n,
    "contributorAddress": "0x1234",
    "decayMultiplier": 0.9999968290208017,
    "lineId": "line2",
    "score": "50",
    "translatedLineId": "translatedLine4",
    "translatedText": "Hello, Bob!",
  },
  {
    "contributedAtMs": 1649536100000n,
    "contributorAddress": "0x1234",
    "decayMultiplier": 0,
    "lineId": "line2",
    "score": "2",
    "translatedLineId": "translatedLine7",
    "translatedText": "Hello, Bobby!",
  },
]
`;
